### ISSUE TITLE
 Persistor debug: Error in load: SerializationException: Load: FormatException:

### ISSUE THREAD

**Issue by mendieta** [Cretezy/redux_persist#25]
Hi, 

I have an issue, after saving the data successfully and restarting the app the persistor throws an error while loading the data.

Load: FormatException: Unexpected character (at character 1965)
...P71hvMH0hAclthtxWaj6KBErGuR9ovy6qr_oBUFDuNcp_1Vk-g","isAuthorized":true}}}}

The error i that the json string has one extra `}` at the end, i have inspected the saves json and all the data is correct except for that  `}` at the end.

Any idead of what may be causing this?

Thanks

**Comment by mendieta** [comment]
After more digging i found that the json string is being sent correctly to the FlutterStorage, adding a breakpoint to line 63 of `storage.dart` confirms this. But just after that last save if i inspect the json file stored in the device it has that extra `}`. 

Its the same issue someone talked about in #16 but was closed.

After some more digging i tested with `FlutterSaveLocation.sharedPreferences` and the error does not occur. 

I will keep working using sharedPreferences, but it may be worthwhile too look at that error. 

Thank you

**Comment by PiercarloSlavazza** [comment]
Same error here.

Really, really **critical**, because data gets corrupted in a way which seems totally random, and hardly recoverable. Some times it's just matter of "extra" characters as in the case reported by @mendieta , but in some other cases the json is corrupted "in the middle", leaving with few or null recovery options.

Unofrtunately, I cannot reporduce the issue - it just seem to occur randomly (which, in generl, might suggest that's a multithreading issue, but in case of Dart I am not so sure this correlation might hold).

@mendieta did you had the chance to gain some more insights into this matter?

Switching to `FlutterSaveLocation.sharedPreferences` does not seem to be a suitable workaround given that the shared_preferences pluging home page states: "this plugin must not be used for storing critical data".

I am using:

```
redux: "^3.0.0"
redux_persist_flutter: '^0.6.0'
```

And flutter doctor says:

```
Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel master, v0.9.3-pre.32, on Microsoft Windows [Versione 10.0.17134.285], locale it-IT)
[√] Android toolchain - develop for Android devices (Android SDK 27.0.3)
[√] Android Studio (version 3.1)
[!] IntelliJ IDEA Ultimate Edition (version 2018.2)
    X Flutter plugin not installed; this adds Flutter specific functionality.
    X Dart plugin not installed; this adds Dart specific functionality.
[√] Connected devices (1 available)
```

**Comment by PiercarloSlavazza** [comment]
@Cretezy I beg your pardon if, as a Dart apprentice, I am going to say bullshits, but, I was looking at [this code](https://github.com/Cretezy/redux_persist/blob/master/packages/redux_persist_flutter/lib/src/storage.dart#L63) and I was wondering what might happen if two isolates try to execute it in a concurrent way: could it be possible at all? Because, if possible, I wonder if this situation might be the cause of the matter discussed here and in issue #16  as well. In #16 it is speculated that the root cause of the problem might be the serialization library - but, from what I have observed, it hardly could be so, because I see JSON which is clearly, say, "intertwined" exactly as if the file was at first written by a proccess/isolate and then partly overwritten by anothe process/isolate - in a clearly "thread-unsafe way" (pardon me again for my "Java parlance") - look for instance at this faulty JSON snippet:

```json
		},
		"testBackEndEnabled": false
	}
}: false
}
}
```

**Comment by PiercarloSlavazza** [comment]
> two isolates try to execute it in a concurrent way

... or maybe it is the very same isolate which is processing another event of the event loop, while waiting for the file system... could it be possible?

**Comment by Cretezy** [comment]
@nullp01nt3r would make sense to add a lock there, I agree.

I'm working on a part-rewrite of this package for v1.0 with a much more solid way to handle serialisation and storage (more modular!).

In this case we don't want to run this code in isolates. IO is non-blocking in Dart, so moving it to isolate would use more memory, could cause race conditions, and is definitely not recommended to do.

Back to the OP:  
I have no clue what would cause that. All the JSON is generated using the built-in JSON engine. Maybe try clearing and trying again? Possibly tried to write twice, as mentioned above. Locks will be included in v1.0.

**Comment by PiercarloSlavazza** [comment]
> @nullp01nt3r would make sense to add a lock there, I agree.

@Cretezy I have already forked the project and done a very simple integration with the `synchronized` library: I am testing it within out test team - I will report here whether the patch fixed the problem after having collected enough evidences.

**Comment by Cretezy** [comment]
Sounds good! Let me know of your findings. I will incorporate the ideas in the next version

**Comment by PiercarloSlavazza** [comment]
> Let me know of your findings

After a week or os of usage of the app w/ the "patched" version of the library, the deserialization issue occurred no more. The app is not yet release, we are talking only abount the testers team, but I believe that the issue has been actually fixed because, before the patch, it occurred very often.

@Cretezy I will send a pull request, so that, if you want, you can merge this simple patch in order to fix the issue right now (w/o having to wait for the 1.0 rewrite).

**Comment by Cretezy** [comment]
This was fixed in 0.8.0-rc.0!
